src/algoseek_connector/base.py

```diff
@@ -0,0 +1,1042 @@
+            raise InvalidDataSetName(key) from None
+
+    def _ipython_key_completions_(self):  # pragma: no cover
+        return list(self.__dict__)
+
+
+class DataSetFetcher:
+    """
+    Lightweight representation of a dataset.
+
+    Manages creation of DataSet instances for querying data using SQL and
+    data downloading.
+
+    Methods
+    -------
+    download:
+        Download data files.
+    fetch:
+        Create a DataSet instance.
+
+    """
+
+    def __init__(self, group: DataGroup, name: str):
+        self._group = group
+        self._source = group.source
+        group_name = self.group.description.name
+        self._description = group.source.description_provider.get_dataset_description(
+            group_name, name
+        )
+        self._dataset = None
+
+    @property
+    def description(self) -> DataSetDescription:
+        """Get the dataset name."""
+        return self._description
+
+    @property
+    def group(self) -> DataGroup:
+        """Get the dataset group."""
+        return self._group
+
+    @property
+    def source(self) -> DataSource:
+        """Get the data source."""
+        return self._source
+
+    def download(
+        self,
+        download_path: Path,
+        date: Union[date_like, tuple[date_like, date_like]],
+        symbols: Union[str, list[str]],
+        expiration_date: Union[date_like, tuple[date_like, date_like], None] = None,
+    ):
+        """
+        Download data from the dataset.
+
+        Parameters
+        ----------
+        download_path : pathlib.Path
+            Path to a directory to download dataset files.
+        date : str, datetime.date or tuple
+            Download data in this date range. Dates can be passed as a str with
+            `yyyymmdd` format or as date objects. If a tuple is passed, it is
+            interpreted as a date range and all dates in the closed interval
+            between the two dates are generated. If a single date is passed,
+            download data from this specific date.
+        symbols : str or list[str]
+            Download data associated with these symbols.
+        expiration date : str, datetime.date, tuple or None, default=None
+            Download data with expiration dates in this date range. Dates must
+            be passed using the same format used for the `date` parameter.
+
+        """
+        self.source.client.download(
+            self.description.name, download_path, date, symbols, expiration_date
+        )
+
+    def fetch(self) -> DataSet:
+        """
+        Create a dataset instance.
+
+        DataSet allow to fetch data using SQL-like queries. See
+        :ref:`here <datasets>` for a detailed description on how work with
+        datasets.
+
+        """
+        if self._dataset is None:
+            dataset = DataSet(self.group, self.description)
+            self._dataset = dataset
+        else:
+            dataset = self._dataset
+        return dataset
```

Comment:
Agreed with the code changes. 
Regarding the implementation of `fetch`, it was done in this way to avoid having multiple instances of the same dataset. So if the user calls the fetch method multiple times, the same dataset instance is retrieved.